code/linear_model.py

```diff
@@ -0,0 +1,156 @@
+import project_config
+import numpy as np
+import numpy.linalg as npl
+import matplotlib.pyplot as plt
+import nibabel as nib
+from general_utils import form_cond_filepath, prepare_data_single, prepare_mask
+from stimuli_revised import events2neural_std
+from conv import conv_target_non_target, conv_std
+from gaussian_filter import spatial_smooth
+from matplotlib import colors
+from hypothesis import perform_t_tests
+
+import pdb
+
+def single_subject_linear_model(standard_source_prefix, cond_filepath_prefix, subject_num, task_num):
+
+  data = prepare_data_single(subject_num, task_num, True, standard_source_prefix)
+
+  n_trs = data.shape[-1] + 5
+
+  cond_filename_003 = form_cond_filepath(subject_num, task_num, "003", cond_filepath_prefix)
+  cond_filename_005 = form_cond_filepath(subject_num, task_num, "005", cond_filepath_prefix)
+  cond_filename_001 = form_cond_filepath(subject_num, task_num, "001", cond_filepath_prefix)
+  cond_filename_004 = form_cond_filepath(subject_num, task_num, "004", cond_filepath_prefix)
+
+  # TODO: put cond007 back to 003
+  cond_filename_007 = form_cond_filepath(subject_num, task_num, "007", cond_filepath_prefix)
+
+  target_convolved, nontarget_convolved = conv_target_non_target(n_trs, cond_filename_003, cond_filename_007, TR, tr_divs = 100.0)
+  target_convolved, nontarget_convolved = target_convolved[5:], nontarget_convolved[5:]
+
+  block_regressor = events2neural_std(cond_filename_005, TR, n_trs)[5:]
+
+  block_start_cues = conv_std(n_trs, cond_filename_001, TR)[5:]
+  block_end_cues = conv_std(n_trs, cond_filename_004, TR)[5:]
+
+  linear_drift = np.linspace(-1, 1, n_trs)
+  qudratic_drift = linear_drift ** 2
+  qudratic_drift -= np.mean(qudratic_drift)
+
+  linear_drift = linear_drift[5:]
+  qudratic_drift = qudratic_drift[5:]
+
+  in_brain_mask, in_brain_tcs = prepare_mask(data, 5000)
+
+  Y = in_brain_tcs.T
+  Y_demeaned = Y - np.mean(Y, axis=1).reshape([-1, 1])
+  unscaled_cov = Y_demeaned.dot(Y_demeaned.T)
+  U, S, V = npl.svd(unscaled_cov)
+
+  X = np.ones((n_trs - 5, 12))
+  X[:, 0] = target_convolved
+  X[:, 1] = nontarget_convolved
+  X[:, 2] = block_regressor
+  X[:, 3] = block_start_cues
+  X[:, 4] = block_end_cues
+  X[:, 5] = linear_drift
+  X[:, 6] = qudratic_drift
+  X[:, 7] = U[:,0]
+  X[:, 8] = U[:,1]
+  X[:, 9] = U[:,2]
+  X[:, 10] = U[:,3]
+  # 11th column is the intercept
+
+  B = npl.pinv(X).dot(Y)
+
+  rs_squared = []
+  for i in range(Y.shape[-1]):
+    r_squared = 1 - np.sum((Y[:,i] - X.dot(B[:,i]))**2) * 1.0 / np.sum((Y[:,i] - np.mean(Y[:,i])) ** 2)
+    rs_squared.append(r_squared)
+
+  print "mean R squared across all voxels is " + str(np.mean(rs_squared))
+
+  pad_thickness = 3
+  fwhm = 4
+  t_test_target_beta = 0
+
+  p_value = perform_t_tests(X, B, Y, t_test_target_beta)
+
+  p_vols = np.zeros((data.shape[0:-1]))
+  p_vols[in_brain_mask] = p_value
+  p_vols_smooth = spatial_smooth(p_vols.reshape(p_vols.shape + (1,)), in_brain_mask, pad_thickness, fwhm)
```

Comment:
You are right. We need to state the assumption that voxels are similar functionally compared to those nearby. I think smoothing reduces certain noise that is not detectable by looking at say, first principal components. 

One thing I think I'll change is to smooth before we fit the linear model rather than after.